### PR TITLE
[12.0][MIG] project task material stock

### PR DESCRIPTION
**Pull request by ValentinVinagre**
Migration project task material stock.

### CODE REVIEW

File: project_task_material_stock/models/project.py [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+        comodel_name='stock.location',
+        string='Source Location',
+        index=True,
+        help='Default location from which materials are consumed.',
+    )
+    location_dest_id = fields.Many2one(
+        comodel_name='stock.location',
+        string='Destination Location',
+        index=True,
+        help='Default location to which materials are consumed.',
```
Comment (nikul-serpentcs) [inline project_task_material_stock/models/project.py:21]: `Default Destination location to which materials are consumed`

File: project_task_material_stock/models/project.py [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+
+class ProjectProject(models.Model):
+    _inherit = 'project.project'
+
+    location_source_id = fields.Many2one(
+        comodel_name='stock.location',
+        string='Source Location',
+        index=True,
+        help='Default location from which materials are consumed.',
```
Comment (nikul-serpentcs) [inline project_task_material_stock/models/project.py:15]: `Default Source location to which materials are consumed`

**Review by nikul-serpentcs — COMMENTED** [review]
nitpicking

**Review by HaraldPanten — APPROVED** [review]
Use cases that I've tried:

1- Consume materials with stock.

2- Consume materials without enough stock. Changing the stock quantity, and rechecking availability.

3- Consume materials with serial number.

4- Consume material with lot number.

5- Consume material with a reordering rule. After checking availability and running the scheduler, the PO is created properly.

6- Consume materials with a MTO Route. MTO Route is not running after checking availability and even running the scheduler manually. Probably it's because the picking is not coming from a SO? Even though, it's an improvement and not a must.

7- Consuming materials with a standard cost, real cost and average cost. All cases of use are considered in analytical accounting cost/revenue.

8- The cost of consuming materials is compatible with timesheet.

Functional review OK 👍

**Review by osimallen — APPROVED** [review]
Reviewed code and did functional testing. Looks good.

**Comment by osimallen** [comment]
@nikul-serpentcs Do you have a timeline for fixing the runbot/travis errors so this PR can be merged? Thanks!

**Comment by nikul-serpentcs** [comment]
> @nikul-serpentcs Do you have a timeline for fixing the runbot/travis errors so this PR can be merged? Thanks!

Yes @osimallen I will check ASAP

**Comment by osimallen** [comment]
@ValentinVinagre Will you be able to resolve the travis/runbot errors so this PR can be merged? I mistakenly thought @nikul-serpentcs was the original creator when I sent my previous note.Thanks!

File: project_task_material_stock/models/project_task.py [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+            if 'stage_id' in vals or 'material_ids' in vals:
+                if task.consume_material:
+                    todo_lines = task.material_ids.filtered(
+                        lambda m: not m.stock_move_id
+                    )
+                    if todo_lines:
+                        todo_lines.create_stock_move()
+                        todo_lines.create_analytic_line()
+                else:
+                    if task.unlink_stock_move() and task.material_ids.mapped(
```
Comment (nikul-serpentcs) [inline project_task_material_stock/models/project_task.py]:
```suggestion 
                     if task.unlink_stock_move() and task.material_ids.mapped(

```

File: project_task_material_stock/models/project_task.py [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+                picking_id.unlink()
+
+    @api.multi
+    def _update_unit_amount(self):
+        # The analytical amount is updated with the value of the
+        # stock movement, because if the product has a tracking by
+        # lot / serial number, the cost when creating the
+        # analytical line is not correct.
+        for sel in self.filtered(
+           lambda x: x.stock_move_id.state == 'done' and
```
Comment (nikul-serpentcs) [inline project_task_material_stock/models/project_task.py]:
```suggestion
               lambda x: x.stock_move_id.state == 'done' and
```

File: project_task_material_stock/models/project_task.py [hunk 5]
```diff
@@ -0,0 +1,287 @@
...
+        for task in self:
+            if 'stage_id' in vals or 'material_ids' in vals:
+                if task.consume_material:
+                    todo_lines = task.material_ids.filtered(
+                        lambda m: not m.stock_move_id
+                    )
+                    if todo_lines:
+                        todo_lines.create_stock_move()
+                        todo_lines.create_analytic_line()
+                else:
```
Comment (nikul-serpentcs) [inline project_task_material_stock/models/project_task.py:117]:
```suggestion
                else:
                    if task.unlink_stock_move() and task.material_ids.mapped(
                            'analytic_line_id'):

File: project_task_material_stock/models/project_task.py [hunk 6]
```diff
@@ -0,0 +1,287 @@
...
+               picking_id.state == 'draft':
+                picking_id.unlink()
+
+    @api.multi
+    def _update_unit_amount(self):
+        # The analytical amount is updated with the value of the
+        # stock movement, because if the product has a tracking by
+        # lot / serial number, the cost when creating the
+        # analytical line is not correct.
+        for sel in self.filtered(
```
Comment (nikul-serpentcs) [inline project_task_material_stock/models/project_task.py:273]:
```suggestion
        for sel in self.filtered(lambda x: x.stock_move_id.state == 'done' and
                x.analytic_line_id.amount != x.stock_move_id.value):
            sel.analytic_line_id.amount = sel.stock_move_id.value
```

**Comment by nikul-serpentcs** [comment]
@ValentinVinagre Keep suggestion, Travis will be Happy

@osimallen

**Comment by ValentinVinagre** [comment]
Hi,
@nikul-serpentcs I can't change this PR, because I don't have permissions to write in the original repository. 

How can solve this?

**Comment by pedrobaeza** [comment]
Open a new PR taking this one as base.

**Comment by rafaelbn** [comment]
@ValentinVinagre open a new PR taking this one as base and close this one thanks

**Comment by rafaelbn** [comment]
Hi @ValentinVinagre and all, this PR has bots :red_circle: and as described here https://github.com/OCA/project/pull/545#issuecomment-522950478 this should be closed and opened a new one because original author cannot finish it. So closing.

cc @HaraldPanten

**Pull request closed without merging**